gst/nnstreamer/tensor_query/tensor_query_common.c

```diff
@@ -89,3 +140,397 @@ nnstreamer_query_connect (uint64_t id, const char *ip, uint32_t port,
+      if (!query_tcp_send (conn->socket, (uint8_t *) & data->cmd,
+              sizeof (TensorQueryCommand), conn->cancellable)) {
+        nns_loge ("Failed to send to socket");
+        return -EIO;
+      }
+      if (data->cmd == _TENSOR_QUERY_CMD_TRANSFER_DATA) {
+        /* send data */
+        if (!query_tcp_send (conn->socket, (uint8_t *) data->data.data,
+                data->data.size, conn->cancellable)) {
+          nns_loge ("Failed to send to socket");
+          return -EIO;
+        }
+      } else {
+        /* send data_info */
+        if (!query_tcp_send (conn->socket, (uint8_t *) & data->data_info,
+                sizeof (TensorQueryDataInfo), conn->cancellable)) {
+          nns_loge ("Failed to send to socket");
+          return -EIO;
+        }
+      }
+      break;
+    default:
+      /* NYI */
+      return -ENOSYS;
+  }
+  return 0;
+}
+
+/**
+ * @brief close connection with corresponding id.
+ * @return 0 if OK, negative value if error
+ */
+int
+nnstreamer_query_close (int64_t id)
+{
+  TensorQueryConnection *conn = g_hash_table_lookup (table, &id);
+  if (!conn) {
+    nns_loge ("ID does not exist in the table");
+    return -EINVAL;
+  }
+  switch (conn->protocol) {
+    case _TENSOR_QUERY_PROTOCOL_TCP:
+    {
+      GError *err;
+      if (!g_socket_close (conn->socket, &err)) {
+        nns_loge ("Failed to close socket: %s", err->message);
+        g_error_free (err);
+        return -EIO;
+      }
+      g_object_unref (conn->socket);
+      g_object_unref (conn->cancellable);
+
+      G_LOCK (splock);
+      g_hash_table_remove (table, &id);
+      G_UNLOCK (splock);
+    }
+      break;
+    default:
+      /* NYI */
+      return -ENOSYS;
+  }
+  return 0;
+}
+
+/**
+ * 
+ * @brief return initialized server_data
+ * @param src_info tensors info shared with serversrc element
+ * @return void* casted type of TensorQueryServerData*
+ */
+void *
+nnstreamer_query_server_data_init (void)
+{
+  TensorQueryServerData *sdata = g_new (TensorQueryServerData, 1);
+  sdata->cancellable = g_cancellable_new ();
+  sdata->protocol = _TENSOR_QUERY_PROTOCOL_END;
+  /* init union */
+  memset (sdata->_dummy, 0, sizeof (sdata->_dummy));
+  return (void *) sdata;
+}
+
+/**
+ * @brief free server data
+ */
+void
+nnstreamer_query_server_data_free (void *server_data)
+{
+  TensorQueryServerData *sdata = (TensorQueryServerData *) server_data;
+  GError *err;
+  if (!sdata)
+    return;
+
+  switch (sdata->protocol) {
+    case _TENSOR_QUERY_PROTOCOL_TCP:
+      if (sdata->server_socket) {
+        if (!g_socket_close (sdata->server_socket, &err)) {
+          nns_logw ("Failed to close socket: %s", err->message);
+          g_error_free (err);
+        }
+        g_object_unref (sdata->server_socket);
+      }
+      g_object_unref (sdata->cancellable);
+      break;
+    default:
+      /* NYI */
+      nns_logw ("Invalid protocol");
+      break;
+  }
+  g_free (sdata->server_host);
+  g_free (sdata);
```

Comment:
Let's free hash table data and destroy hash table here.

Reply:
I removed hash table from common which is only dependent to server. It would belong to query_server.c